homeassistant/components/frank_energie/coordinator.py

```diff
@@ -0,0 +1,146 @@
+"""Coordinator implementation for Frank Energie integration."""
+from __future__ import annotations
+
+from datetime import date, timedelta
+import logging
+from typing import TypedDict
+
+from python_frank_energie import FrankEnergie
+from python_frank_energie.exceptions import AuthException, RequestException
+from python_frank_energie.models import Invoices, MarketPrices, MonthSummary, PriceData
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt as dt_util
+
+from .const import CONF_AUTH_TOKEN, CONF_REFRESH_TOKEN, DeviceResponseEntry
+
+LOGGER = logging.getLogger(__name__)
+
+
+class FrankEnergieData(TypedDict):
+    """Frank Energie data."""
+
+    DATA_ELECTRICITY: PriceData
+    DATA_GAS: PriceData
+    DATA_MONTH_SUMMARY: MonthSummary | None
+    DATA_INVOICES: Invoices | None
+
+
+class FrankEnergieCoordinator(DataUpdateCoordinator[DeviceResponseEntry]):
+    """Get the latest data and update the states."""
+
+    api: FrankEnergie
+
+    def __init__(
+        self, hass: HomeAssistant, entry: ConfigEntry, api: FrankEnergie
+    ) -> None:
+        """Initialize the data object."""
+        self.hass = hass
+        self.entry = entry
+        self.api = api
+
+        super().__init__(
+            hass,
+            LOGGER,
+            name="Frank Energie coordinator",
+            update_interval=timedelta(minutes=60),
+        )
+
+    async def _async_update_data(self) -> DeviceResponseEntry:
+        """Get the latest data from Frank Energie."""
+        LOGGER.debug("Fetching Frank Energie data")
+
+        # We request data for today up until the day after tomorrow.
+        # This is to ensure we always request all available data.
+        today = dt_util.utcnow().date()
+        tomorrow = today + timedelta(days=1)
+        day_after_tomorrow = today + timedelta(days=2)
+
+        # Fetch data for today and tomorrow separately,
+        # because the gas prices response only contains data for the first day of the query
+        try:
+            prices_today = await self.__fetch_prices_with_fallback(today, tomorrow)
+            prices_tomorrow = await self.__fetch_prices_with_fallback(
+                tomorrow, day_after_tomorrow
+            )
+
+            data_month_summary = (
+                await self.api.month_summary() if self.api.is_authenticated else None
+            )
+            data_invoices = (
+                await self.api.invoices() if self.api.is_authenticated else None
+            )
+        except UpdateFailed as err:
```

Comment:
What is raising `UpdateFailed` here?

Reply:
Ah this was when we did the API calls directly from HA. Removed!